### ISSUE TITLE
time series prediction 

### ISSUE THREAD

**Issue by JiweiZh** [MaukWM/energy_prediction#1]
Hi, I am currently doing a similar time series forecasting for numbers of calls with seq2seq model. Now I meet a problem which is that when I want to predict like 30 days target's values in the future, the results will not perform any trend of series. 

The code of the model is like this: 
```
# train model
latent_dim = 128 # LSTM hidden units
dropout = .20 

encoder_inputs = Input(shape=(None, n_inputs), name='encoder_inputs') 

encoder_lstm = LSTM(latent_dim, dropout=dropout, return_sequences=False, return_state=True, name='encoder_lstm')

_, state_h, state_c = encoder_lstm(encoder_inputs)

encoder_states = [state_h, state_c]


decoder_inputs = Input(shape=(None, n_features), name='decoder_inputs') 


decoder_lstm = LSTM(latent_dim, dropout=dropout, return_sequences=True, return_state=True, name='decoder_lstm')
decoder_outputs, _, _ = decoder_lstm(decoder_inputs, initial_state=encoder_states)

decoder_dense = (Dense(n_outputs, name='decoder_dense', activation='linear')) 
decoder_outputs = decoder_dense(decoder_outputs)

model = Model([encoder_inputs, decoder_inputs], decoder_outputs)

model.compile(COCOB(), loss=ks.losses.mean_squared_error)

history = model.fit([encoder_input_data, decoder_input_data], decoder_target_data,
                     batch_size=16,  epochs=100,
                     validation_split=0.2, shuffle=False)
```
```

# define inference encoder
encoder_model = Model(encoder_inputs, encoder_states)

decoder_state_input_h = Input(shape=(latent_dim,))
decoder_state_input_c = Input(shape=(latent_dim,))
decoder_states_inputs = [decoder_state_input_h, decoder_state_input_c]  

decoder_outputs, state_h, state_c = decoder_lstm(decoder_inputs, initial_state=decoder_states_inputs)

decoder_states = [state_h, state_c] 

decoder_outputs = decoder_dense(decoder_outputs) 
decoder_model = Model([decoder_inputs] + decoder_states_inputs,
                      [decoder_outputs] + decoder_states)
```
```
# decoded sequence 
n_test_samples = encoder_input_data_test.shape[0]
print('Total number of test smaples:', n_test_samples)

n_seq_out = 1  #length of days for predicting, objective is to predict 3 months in the future
all_out = []

for i in range(n_test_samples):
    enc_outs = encoder_model.predict(encoder_input_data_test[i:i+1,:,:])
    
    target_seq = np.zeros((1,1,n_features))
    
    target_seq[0,:,:] = encoder_input_data_test[i,-1:,:]
    
    decoded_seq = np.zeros((1,n_seq_out,n_outputs))

    for j in range(n_seq_out):
        output, h, c = decoder_model.predict([target_seq]+enc_outs)
        print('out', output)
        print('tar', target_seq)
        decoded_seq[0,j,0] = output[0,0,0] 
        
        target_seq = np.zeros((1,1,n_features))
        target_seq[0,0,0] = output[0,0,0]
        target_seq[0,0,1:] = decoder_input_data_test[j+1,0,1:]
        enc_outs = [h, c]  
        
    all_out.append(decoded_seq)
```
My data input of encoder has the size (n_samples, steps=84 days, n_features), and the data of decoder input has the size(n_samples, steps=1 days, n_features) and the size of target data of decoder is (n_sample, steps=1, 1). Each sample of input data of encoder has one-day lag,  and decoder input data is for realizing the teacher forcing.  

Like if I just predict one day after, the predicted value is acceptable but not very precise. However, when I predict like 30 days after, the values of the predicted series will only increase or decrease unstoppable. Do you have any suggestion on this? 

Thanks a lot!

**Comment by MaukWM** [comment]
Hi redtea0, sorry for the late response.

I'm not sure how long the output sequence is you are training your model on. Also, how quickly does the error build? If I know what prediction horizon you're training your models with and have some visuals of your data I can maybe help you a bit better. I trained my models on 24 hours of input, to predict 24 hours ahead. The models perform quite well as long as they must predict within this 24-hour period, as you can see in this example prediction here:
![s2s_1_day](https://user-images.githubusercontent.com/22435746/59968880-e2ae5700-9541-11e9-9483-2392903b8437.png)

However when this prediction horizon grows larger, the models lose their predictive power (in my case), as you can see in the following image where it attempts to predict 7 days ahead, while only being trained to predict 1 day ahead.
![s2s_7days](https://user-images.githubusercontent.com/22435746/59968811-518ab080-9540-11e9-95c6-3e26552fcfc9.png)

I shortly trained a model with 2 days of input and 6 days of output. The result of a prediction look like this: 
![s2s6](https://user-images.githubusercontent.com/22435746/59969156-cf05ef00-9547-11e9-9631-4687c91b0c41.png)
The result are okay.
Now if we let it predict 15 days ahead let's see what it looks like:
![s2s_15](https://user-images.githubusercontent.com/22435746/59969168-1d1af280-9548-11e9-9268-1b78b2ef56eb.png)
As you can see the model at least does not stagnate anymore, which is quite nice! Important to note is the seq2seq model does not seem to make very different predictions, it chooses a pattern an sticks quite closely to it. It does not copy its predictions fully tho, as you can see the peaks of the predictions for each day increase over time. But it is clear that the network realizes that a simple cyclical pattern can already reduce the error by a lot (in my case the data is much alike a sine wave).

Something else you can also try is adding attention to your model. For my model, which is trained to predict just 24 hours, ahead this is the result:
![s2s_attention](https://user-images.githubusercontent.com/22435746/59968842-e8576d00-9540-11e9-8ea2-746d86ba5716.png)
As you can see the result is better than with a pure sequence to sequence model trained on just one-day ahead predictions. But if you pay close attention (no pun intended), to the pattern of these predictions, you can see they are purely cyclical. If we look at what the network pays attention to, then you can see it is just repeating its previous prediction over and over again. The cyclical pattern is clear from this image.
![7_day_attention](https://user-images.githubusercontent.com/22435746/59968858-379d9d80-9541-11e9-9ec3-45a17b227d77.png)
The x-axis represents the current timestep it is predicting, and the y-axis represents each computed intermediate state of the encoder. The more bright a square is, the more attention it is paying to that specific intermediate state from the encoder.

So we run into the same problem as a seq2seq model trained for predicting with a 6-day prediction horizon.

With attention, states within the decoder are also completely limited to states computed within the encoder. The new state at each step is computed from a weighted sum of all intermediate states from the encoder. I have no proof but this could hinder you from making predictions outside of your trained prediction horizon. So attention might not be the solution you are looking for, so if you plan on doing this approach, approach with care, but feel free to play around with attention as it might be a useful addition to your models.

I hope this information is relevant to your problem and can help you. If you have any interesting findings I would be very much interested in knowing!

**Comment by JiweiZh** [comment]
Hi, thx for your clear explanation. 
For my job, I need to predict the day of 3 months later, so I thought that I have two ways:
   1. The model is trained on 91 days of input, and predict one day after 90 days. For example, the first sample starts at 2012-01-01 and ends at 2012-04-01 (input encoder contains 91 days), and then predict one day 2012-07-01( as decoder target data), each sample has one day lag, the second sample starts at 2012-01-02 and ends at 2012-04-02 in order to predict the day of2012-07-02.
 But at the same time, I have a decoder input data for teacher forcing which have one day lag before of decoder target data. 

![image](https://user-images.githubusercontent.com/16006025/60027994-d5f34580-969e-11e9-8923-cb7a08d5b406.png)
![image](https://user-images.githubusercontent.com/16006025/60028030-e5728e80-969e-11e9-8b9b-4db8dcdbb7dc.png)
![image](https://user-images.githubusercontent.com/16006025/60028064-f91df500-969e-11e9-9281-033e94d975ba.png)

I train the model on 2101 days and test on 515 days, but the result is awful. I don't know why and how to solve that.

2. The second way, I tried to train the model on 91 days, and then predict 91 days ahead. 

![image](https://user-images.githubusercontent.com/16006025/60029971-7008bd00-96a2-11e9-913c-f816d382a3a1.png)
![image](https://user-images.githubusercontent.com/16006025/60028608-ebb53a80-969f-11e9-8129-a60d2358b675.png)
![image](https://user-images.githubusercontent.com/16006025/60029982-7434da80-96a2-11e9-97f7-72f00dc0c478.png)

The results are more acceptable, but not enough. 

I began to doubt my features. Because each of my sample of encoder input contains 91 days timesteps and 19 features. This first column of 19 features is the Number that I want to predict, which is also the target of the decoder. And other 18 features are external features (Year, Month, Day, Weekday, holiday, before holiday....), after the normalization and separate sliding window, I put them together as 19 features into the model Seq2seq. I'm not sure if I can do this or not. 

If there is something I didn't explain clearly, please let me know.

**Comment by MaukWM** [comment]
Hi, I have a few questions for you. First, what exactly are the contents of the graphs you attached? And second, why exactly do you want to predict the value exactly 3 months later? If I know the context I might be able to help better or suggest some other methods.

If I take a look at the first 2 graphs of both methods the results don't seem that bad. What kind of error metrics do you use? If visual verification is all you're using I really recommend you perform some predictions on the trained models and evaluate them according to some error measures. You can try calculating the [RMSE and NRMSE](https://en.wikipedia.org/wiki/Root-mean-square_deviation) (NRMSD) for example over the entire testing data. This could give you a more formal evaluation of the performance of the models.

You also have to keep in mind that predicting at least 3 months ahead can be very difficult (3 months is a very long time, a lot can happen that can make your prediction very off!), although I am not fully aware of the context of the problem. But you might be trying to solve a problem that is too difficult to solve.

Also are the 18 external features all related to the time and seasonality?

**Comment by JiweiZh** [comment]
Hi, thanks again for your answer ~~ 

Firstly the graphs show the prediction value and true value for this regression question.  The third graph of both two methods is a global visualisation(515 days), the first graph of both is the first 100 days and the second graph of both is the last 100 days. All of this is to compare the true value and predict value. 
  
Secondly, 3 months is the purpose of my work, because the company needs to know exactly 3 months before in order to well budget the people, so I can't change it.  

For evaluating my results, I use SMAPE, MAE and RMSE, for my benchmark made by Random Forest, I have already 7.5% errors for SMAPE, 120 for MAE and 180 for RMSE. However, I need a lot better result than these, so that's why I tried the deep learning with seq2seq model. Furthermore, I looked also the graph, because sometimes the metrics are not too bad, but the actual graph cannot catch the peak or the trend.  

For the features, there exist some internal features related to the time like (Year, Month, Day, WeekDay..), for the features like year, month, or day, I calculate the sin and cos values for them. 
And other external features (Holiday, the day before the holiday, group of days...), I only take the value of themselves (0 or 1). 
For all of these features, after the processing and engineering of the features, I put them directly into the model, that's where I don't know it's correct or not. I thought the model will consider each of the feature as a time series? And for the decoder input data, I'm not sure if the input needs to contain all the features or only target values.

**Comment by MaukWM** [comment]
For teacher guided training, the decoder must receive the _target_ value as input from the previous timestep. In the rough diagram of a seq2seq model below you can see that within the decoder, the previous output is fed back into the network, so when performing guided training, instead of feeding the previous output, you feed it the actual target value of the previous timestep.

![Seq2seq_diagram-1](https://user-images.githubusercontent.com/22435746/60084017-07b7eb00-9737-11e9-9cc3-c030a496d1c2.png)

No model is perfect, we can only hope to produce the best model possible. What you can try to do to improve accuracy (or results):

1. Add more data, but I believe if you had access to more you would have used it.

2. Enhance your current model. You could try adding stuff like Dropout, Regularization or Batch Normalization. You can also try adding bidirectional operations or adding attention (Not sure how much this would help though. From experience attention is more useful for chaotic predictions, but your data has a pretty straightforward pattern)

3. Changing the value you try to predict. Does your company require you to predict the load for each individual day 3 months from now? A potentially easier problem is to predict the total amount for an entire week in one go instead of each day of the week. So you try to predict the value of the first week 3 months from now instead of the first day 3 months from now.
